### ISSUE TITLE
Tax calculation rounded and wrong with All set to exclusive Tax + tax per product + several products

### ISSUE THREAD

**Issue by rk-mxp** [magento/magento2#33352]
### Preconditions (*)

1. Magento 2.4.2-p1

### Steps to reproduce (*)

1. Set every configuration in Sales / Tax to Excluding Tax (Calucation Settings: Catalog Prices, Shipping PRices, Apply Discount on Prices, All Display Settings)
2. Set "The Tax Calculation Method Based On" to "Unit Price"
3. Add a product to the system and add Customer Group Price Configuration to it, so more products cost less, e.g. it gets cheaper if you buy 100+. (not sure if this is required)
4. The product in question was also a configurable product (might be relevant as well).
5. Tax ist added to shipping costs as well (might be irrelevant)
6. Add a discount of 3% to all products as a shopping cart rule (In our case starting at a minimum order amount)
7. Add these products in high quantites to the shopping cart, e.g., 200 (very important)

### Expected result (*)

1. Calculated tax should not be rounded off.

### Actual result (*)

1. Calculated tax is rounded, see screenshot with an example
![Magento_tax_calculation_bug](https://user-images.githubusercontent.com/35734421/123396626-d1213f00-d5a1-11eb-8ffe-d610241db5da.jpg)

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Hotel** [comment]
Hi @Barakur,

We have tried to reproduce this issue in Magento 2.4.2-p1 branch. It seems that tax calculation is working fine. We have tried it with adding cart rule with 3% discount as specified in the screenshot. Please find below screenshot for reference:

![image](https://user-images.githubusercontent.com/51681618/124271799-8718fa00-db5b-11eb-83cb-bb4a5d6706df.png)


The tax is calculated after deducting discount amount. But the tax amount rounding off, this seems to us an issue because it should not rounding off. 

So please let us know if you want confirm this issue for rounding off the tax amount.

Thanks

**Comment by Barakur** [comment]
Yes, rounding the tax shouldn't happen. Some more screenshots for reference (from the same order, all items in the order as well as the summary):

![Magento_tax_calculation_bug2](https://user-images.githubusercontent.com/35734421/124422108-104e4d80-dd63-11eb-851d-e732bb8804c3.jpg)
![Magento_tax_calculation_bug3](https://user-images.githubusercontent.com/35734421/124422113-12b0a780-dd63-11eb-93c7-78e21a4b343a.jpg)

Two items have correct tax calculated (without rounding): Those with an amount of only 1. The others are rounded. If you caclulated the tax for the first position:
320.00 - 9,60 = 310,40 (taxable amount)
310,40 * 19% = 58,976 (which gets rounded down(!) to 58)
The rounding is really strange since it isn't round, floor or ceil apparently.

But generally tax should not be rounded to full euro.

**Comment by engcom-Hotel** [comment]
Hi @Barakur,

We are confirming this issue for rounding off the tax, and functionality of tax calculation is working fine. And also we are updating the main issue description because tax calculation is working fine.

Thanks

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Hotel 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42874` were created

**Issue Available**: @engcom-Hotel, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by alex-james** [comment]
In Magento\Tax\Model\Sales\Total\Quote\Subtotal::collect() when Magento\Tax\Model\TaxCalculation::calculateTax() is called, the third argument ($round) is not specified and therefore the default of true is used, this results in rounding occurring in Magento\Tax\Model\Calculation\UnitBaseCalculator::calculateWithTaxNotInPrice() which means the tax calculation is something like:

ROUND(UNIT PRICE * DISCOUNT, 2) * Qty

Adding a plugin to set $round to false results in the following (correct) tax calculation:

UNIT PRICE * DISCOUNT * Qty

**Comment by artskirk** [comment]
@magento I am working on this

**Comment by engcom-Delta** [comment]
Hi @rk-mxp ,

Thanks for your reporting and collaboration.
We have verified the issue in latest 2.4-develop instance and the issue is reproducible.
Kindly refer the screenshots.

**Steps to reproduce**
1. Set every configuration in Sales / Tax to Excluding Tax (Calucation Settings: Catalog Prices, Shipping PRices, Apply Discount on Prices, All Display Settings)
2. Set "The Tax Calculation Method Based On" to "Unit Price"
3. Add a discount of 3% to all products as a shopping cart rule
4. Add two products with high quantity. e.g. 200
5. Apply coupon and place order successfully.
6. Observe tax is rounded off
![image](https://github.com/user-attachments/assets/0435f619-b5b0-4955-affc-47256deeb89b)

Thanks.
